### PR TITLE
[2.7] bpo-28143: ASDL compatibility with Python 3 system interpreter

### PR DESCRIPTION
**Pull request by malthe**
https://bugs.python.org/issue28143

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Parser/spark.py [hunk 1]
```diff
@@ -660,8 +665,8 @@ def ambiguity(self, rules):
             sortlist.append((len(rhs), name))
             name2index[name] = i
         sortlist.sort()
-        list = map(lambda (a,b): b, sortlist)
-        return rules[name2index[self.resolve(list)]]
+        mapped = list(map(lambda ab: ab[1], sortlist))
```
Comment (serhiy-storchaka) [inline Parser/spark.py]: mapped = [b for a, b in sortlist]

File: Parser/spark.py [hunk 2]
```diff
@@ -825,15 +830,15 @@ def resolve(self, list):
...
+        print('set', i)
         for item in sets[i]:
-            print '\t', item
+            print('\t', item)
             for (lhs, rhs), pos in states[item[0]].items:
-                print '\t\t', lhs, '::=',
-                print string.join(rhs[:pos]),
-                print '.',
-                print string.join(rhs[pos:])
+                 print('\t\t', lhs, '::=', end='')
```
Comment (serhiy-storchaka) [inline Parser/spark.py]: `end=' '`

But I think it would be better to use just a single multiline `print()`

File: Parser/spark.py [hunk 3]
```diff
@@ -418,10 +423,10 @@ def makeState(self, state, sym):
         #  need to know the entire set of predicted nonterminals
         #  to do this without accidentally duplicating states.
         #
-        core = predicted.keys()
+        core = list(predicted.keys())
```
Comment (serhiy-storchaka) [inline Parser/spark.py]: I think this can be written just as
`tcore = tuple(sorted(predicted))`

File: Parser/spark.py [hunk 4]
```diff
@@ -312,12 +313,15 @@ def parse(self, tokens):
             self.states = { 0: self.makeState0() }
             self.makeState(0, self._BOF)
 
-        for i in xrange(len(tokens)):
+        i = 0
+        while i < len(tokens):
```
Comment (serhiy-storchaka) [inline Parser/spark.py]: You could use `enummerate()`.

File: Parser/spark.py [hunk 5]
```diff
@@ -30,7 +32,7 @@ def _namelist(instance):
         for b in c.__bases__:
             classlist.append(b)
         for name in c.__dict__.keys():
```
Comment (serhiy-storchaka) [inline Parser/spark.py]: `.keys()` is redundant.

File: Parser/asdl_c.py [hunk 6]
```diff
@@ -1194,7 +1194,7 @@ def main(srcfile):
         sys.exit(1)
     if INC_DIR:
         p = "%s/%s-ast.h" % (INC_DIR, mod.name)
-        f = open(p, "wb")
+        f = open(p, "w")
```
Comment (serhiy-storchaka) [inline Parser/asdl_c.py]: The "wb" mode is used for disallowing newlines translation. In Python 3 you should use `newline='\n'`.

**Comment by serhiy-storchaka** [comment]
Currently this doesn't work neither with Python 2 nor with Python 3.

File: Parser/asdl_c.py (hunk 6 above)
Comment (malthe) [reply Parser/asdl_c.py]: This is Python 2.7.

File: Parser/spark.py (hunk 5 above)
Comment (malthe) [reply Parser/spark.py]: Yep. But this is not a rewrite of the `Parser/spark.py` module. I did not touch this line.

File: Parser/spark.py (hunk 4 above)
Comment (malthe) [reply Parser/spark.py]: Fixed in 34ea1cf9f7be27c12156d6c9e6c19984068e44a9.

File: Parser/spark.py (hunk 3 above)
Comment (malthe) [reply Parser/spark.py]: Thanks. Fixed in aef0c885e404c0165a89e2cff1920d6fd4440705.

File: Parser/spark.py (hunk 1 above)
Comment (malthe) [reply Parser/spark.py]: Fixed in ebea5cdeac637c43c87cf03bb513bf9050a81eca.

File: Parser/spark.py (hunk 2 above)
Comment (malthe) [reply Parser/spark.py]: Yeah. My going principle was as few changes as possible – to make it easier to verify correctness.

**Comment by malthe** [comment]
@serhiy-storchaka sorry, I had some trouble applying the original patch so it was only partially complete. It should work now.

File: Parser/asdl_c.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Parser/asdl_c.py]: Then why this is changed?

File: Parser/asdl.py [hunk 7]
```diff
@@ -304,9 +306,9 @@ def visit(self, object, *args):
             return
         try:
             meth(object, *args)
-        except Exception, err:
-            print "Error visiting", repr(object)
-            print err
+        except Exception as err:
```
Comment (serhiy-storchaka) [inline Parser/asdl.py:309]: This syntax is not compatible with Python 2.5. Use `sys.exc_info()` for more compatibility.

File: Parser/spark.py [hunk 8]
```diff
@@ -825,15 +826,15 @@ def resolve(self, list):
...
+        print('set', i)
         for item in sets[i]:
-            print '\t', item
+            print('\t', item)
             for (lhs, rhs), pos in states[item[0]].items:
-                print '\t\t', lhs, '::=',
-                print string.join(rhs[:pos]),
-                print '.',
-                print string.join(rhs[pos:])
+                print('\t\t', lhs, '::=', end=' ')
```
Comment (serhiy-storchaka) [inline Parser/spark.py]: `end=` is not supported in Python 2.5. Use the single `print()`.

File: Parser/spark.py [hunk 9]
```diff
@@ -644,7 +645,7 @@ def buildTree(self, nt, item, tokens, k):
                 attr[i] = self.buildTree(sym, why[0],
                                          tokens, why[1])
                 item, k = self.predecessor(key, why)
-        return self.rule2func[self.new2old[rule]](attr)
+        return self.rule2func[self.new2old[rule]](*attr)
```
Comment (serhiy-storchaka) [inline Parser/spark.py:648]: Shouldn't the call in `deriveEpsilon()` be updated too?

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

**Comment by serhiy-storchaka** [comment]
Python 3 emits a warning:
```
/home/serhiy/py/cpython2.7/Parser/asdl.py:387: ResourceWarning: unclosed file <_io.TextIOWrapper name='./Parser/Python.asdl' mode='r' encoding='UTF-8'>
  buf = open(file).read()
```

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Make the code compatible with Python 2.5, fix the problem with a resource warning and newlines translation.

**Comment by csabella** [comment]
@malthe, please address the code review from @serhiy-storchaka.

File: Parser/asdl_c.py (hunk 6 above)
Comment (malthe) [reply Parser/asdl_c.py]: Addressed now in 8405be4a2e58bc4d06117da84f581e8514da614c.

File: Parser/spark.py (hunk 9 above)
Comment (malthe) [reply Parser/spark.py:648]: I think so yes. Fixed in e14e5361932c181b1c0ab9be7b685fdfbc23f8a8.

File: Parser/spark.py (hunk 8 above)
Comment (malthe) [reply Parser/spark.py]: Fixed in 39e5248e39b35b9da2d57590361eacdf98c6e8e5.

**Comment by malthe** [comment]
@csabella, I've made the necessary changes. I don't know if the continuous integration test actually runs this code. I assume that there isn't anything set up for running the build with a default Python 3 interpreter since this is what was failing.

**Pull request closed without merging**